url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPSystemManager.java

```diff
@@ -0,0 +1,1344 @@
+     * units have been disconnected.
+     */
+    @Override
+    public final synchronized void disconnect() {
+        if (state == SystemState.STOPPED) {
+            return; // Already stopped
+        }
+
+        sysThread.shutdownThread();
+        state = recalcMode();
+    }
+
+    /**
+     * Reload the properties from the default file jOCP.properties. This will
+     * result in all OCP links being temporarily lost.
+     */
+    public final void reloadProperties() {
+        FileInputStream in = null;
+        Properties myProp = new Properties();
+
+        try {
+            in = new FileInputStream("jOCP.properties");
+            myProp.load(in);
+        } catch (FileNotFoundException e) {
+            log.warn("Could not find the properties file \"jOCP.properties\". "
+                    + "Falling back to defaults.", e);
+        } catch (IOException e) {
+            log.error("Error reloading properties file \"jOCP.properties\". "
+                    + "Falling back to defaults.", e);
+        } finally {
+            try {
+                if (in != null) {
+                    in.close();
+                }
+            } catch (IOException e) { // CSIGNORE: EmptyBlock
+            }
+        }
+
+        reloadProperties(myProp);
+    }
+
+    /**
+     * Reload the properties using the specified Properties object. This will
+     * result in all OCP links being temporarily lost.
+     *
+     * @param properties
+     *            The Properties object to use
+     */
+    @Override
+    public final void reloadProperties(final Properties properties) {
+
+        this.prop = properties;
+        init();
+
+    }
+
+    /**
+     * Recalculate the system state, based on the state of each connected link.
+     * The system state is calculated as follows:
+     * <ul>
+     * <li>If the OCP System Manager is not running, then the system state is
+     * set to STOPPED.</li>
+     * <li>If all links are inactive (i.e. in the CONNECTING or DISCONNECTED
+     * states), then the system state is set to CONNECTING.</li>
+     * <li>If all active links are in the SLAVE state, then the system state is
+     * set to MISSING_MASTER.</li>
+     * <li>If exactly one link is in the MASTER state and all other active links
+     * are in the SLAVE state, then the system state is set to MASTER_SLAVE.
+     * </li>
+     * <li>If all active links are in the LOADSHARE state, then the system state
+     * is set to LOADSHARING.</li>
+     * <li>If the links are in any other combination, then the system state is
+     * set to INCONSISTENT.</li>
+     * </ul>
+     * Additionally, if the system state is one of STOPPED, CONNECTING or
+     * INCONSISTENT then an alarm will be raised signalling that the OCP module
+     * cannot take calls. An alarm will also be raised for each link that is in
+     * the CONNECTING or DISCONNECTED state.
+     *
+     * @return the new state of the system
+     */
+    private SystemState recalcMode() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 103 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 509-634 into a separate method.